### PR TITLE
[WIP] k8s 1.25.0-rc.1

### PR DESCRIPTION
**Pull request by soltysh**
Picks: https://docs.google.com/spreadsheets/d/10KYptJkDB1z8_RYCQVBYDjdTlRfyoXILMa0Fg8tnNlY/edit#gid=1852049954
Test tracker: https://docs.google.com/spreadsheets/d/1KuC8QAmQd2LiCMk0EVBxss97V0-UNtFM8vVEqPYWCqs/edit#gid=1312057609

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/patch_etcd3.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	// retry in a loop in the background until we successfully create the client, storing the client or error encountered
+	// the health check is run every second and the result is stored for an immediate healthz response
+
+	clientValue := &atomic.Value{}
+
+	clientErrMsg := &atomic.Value{}
+	clientErrMsg.Store("etcd client connection not yet established")
+
+	go wait.PollUntil(time.Second, func() (bool, error) {
+		client, err := newETCD3Client(c.Transport)
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/patch_etcd3.go:26]: is this creating a new client every second?

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/patch_etcd3.go:26]: ah, no, once it succeeds creating a client it exits. it keeps trying to create a client until it gets it

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/patch_etcd3.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		return true, nil
+	}, wait.NeverStop)
+
+	healthzErrorMessage := &atomic.Value{}
+	healthzErrorMessage.Store("etcd client connection not yet established")
+	go wait.Until(func() {
+		if errMsg := clientErrMsg.Load().(string); len(errMsg) > 0 {
+			healthzErrorMessage.Store(errMsg)
+			return
+		}
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/patch_etcd3.go:42]: ok, this waits until the client has been created, that is when `clientErrMsg` is empty

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/patch_etcd3.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		if err == nil {
+			healthzErrorMessage.Store("")
+			return
+		}
+		healthzErrorMessage.Store(err.Error())
+	}, time.Second, wait.NeverStop)
+
+	return func() error {
+		if errMsg := healthzErrorMessage.Load().(string); len(errMsg) > 0 {
+			return fmt.Errorf(errMsg)
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/patch_etcd3.go:57]: > Until is syntactic sugar on top of JitterUntil with zero jitter factor and with sliding = true (which means the timer for period starts after the f completes).

the wait.Until will drift the probe sampling, if the check start to timeout, we'll be serving stale values IIUIC

**Comment by soltysh** [comment]
Closing in favor of https://github.com/openshift/kubernetes/pull/1360.

**Pull request closed without merging**